### DISCUSSION TITLE
high cpu usage

### DISCUSSION THREAD

**Discussion by franzPilgerstorfer** [nielsbasjes/yauaa#739] [Q&A] [answered]
I have implemented a jsp using yauaa. Therefor I am importing UserAgent and UserAgentAnalyzer, creating necessary objects, calling parse(..) of a given userAgent string and returning either a json value or a simple html page. 

Unfortunatelly if there are multiple calls to this code (Postman runner, 500 calls one after another), we encounter a huge cpu impact. I don' think this is normal but also fail on identifying the problem? 

any help is appreciated!

![image](https://user-images.githubusercontent.com/12933737/190972623-d4d0bc53-4625-4ef1-a90b-ffc533b82821.png)
```

<%@page import="nl.basjes.parse.useragent.UserAgent"%>
<%@page import="nl.basjes.parse.useragent.UserAgentAnalyzer"%>
<%@page language="java" contentType="text/html; charset=ISO-8859-1"	pageEncoding="ISO-8859-1"%>
<%
// stripped out simple html outputs
UserAgentAnalyzer uaa = UserAgentAnalyzer.newBuilder().hideMatcherLoadStats().withCache(10000).build();
UserAgent userAgentData = uaa.parse(userAgentValue); // either from header or from request parameter

// removed some rules
response.setContentType("application/json");
StringBuilder jsonUserAgentData= new StringBuilder();
for (String fieldName : userAgentData.getAvailableFieldNamesSorted()) {
	jsonUserAgentData.append( "\t\t\"" + fieldName + "\" :  \"" + userAgentData.getValue(fieldName) + "\", \n");
}
					
out.println("{\n\"store\": \"" + storeName + "\",\n\"useragentdata\":  {\n" + jsonUserAgentData.toString().substring(0, jsonUserAgentData.length()-3) + " }\n}");
%>
```

**Comment by franzPilgerstorfer** [comment, marked as answer]
Apparently I figured it out on my own. Creating a new UserAgentAnalyzer object for each and every call to this jsp file causes this issue. I've put it into an application scope variable now. Works like a charm.

**Reply by nielsbasjes** [reply]
Good to hear you figured it out.
I double checked and I did put a notice in the documentation about this:
https://yauaa.basjes.nl/using/#using-in-java-applications 

> Please instantiate a new UserAgentAnalyzer as few times as possible because the initialization step for a full UserAgentAnalyzer (i.e. all fields) usually takes something in the range of 2-5 seconds.

**Reply by franzPilgerstorfer** [reply]
Yes, you are perfectly right. This note helped me to figure it out!

**Reply by nielsbasjes** [reply]
I realized the documentation was not clear enough about this so I updated it.
https://yauaa.basjes.nl/using/#using-in-java-applications
